tests/components/unifi/test_controller.py

```diff
@@ -293,33 +321,132 @@ async def test_reset_after_successful_setup(hass, aioclient_mock):
+        await hass.async_block_till_done()
+
+    assert result is False
+
+
+async def test_connection_state_signalling(hass, aioclient_mock, mock_unifi_websocket):
+    """Verify signalling and connection state are properly represented."""
+    config_entry = await setup_unifi_integration(hass, aioclient_mock)
+    controller = hass.data[UNIFI_DOMAIN][config_entry.entry_id]
+
+    event_call = Mock()
+    unsub = async_dispatcher_connect(hass, controller.signal_reachable, event_call)
+
+    assert controller.available is True
+    assert len(event_call.mock_calls) == 0
+
+    mock_unifi_websocket(state=STATE_DISCONNECTED)
+    await hass.async_block_till_done()
+
+    assert controller.available is False
+    assert len(event_call.mock_calls) == 1
+
+    mock_unifi_websocket(state=STATE_RUNNING)
+    await hass.async_block_till_done()
+
+    assert controller.available is True
+    assert len(event_call.mock_calls) == 2
+
+    unsub()
+
+
 async def test_wireless_client_event_calls_update_wireless_devices(
-    hass, aioclient_mock
+    hass, aioclient_mock, mock_unifi_websocket
 ):
     """Call update_wireless_devices method when receiving wireless client event."""
-    config_entry = await setup_unifi_integration(hass, aioclient_mock)
-    controller = hass.data[UNIFI_DOMAIN][config_entry.entry_id]
+    await setup_unifi_integration(hass, aioclient_mock)
 
     with patch(
         "homeassistant.components.unifi.controller.UniFiController.update_wireless_clients",
         return_value=None,
     ) as wireless_clients_mock:
-        controller.api.websocket._data = {
-            "meta": {"rc": "ok", "message": "events"},
-            "data": [
-                {
-                    "datetime": "2020-01-20T19:37:04Z",
-                    "key": aiounifi.events.WIRELESS_CLIENT_CONNECTED,
-                    "msg": "User[11:22:33:44:55:66] has connected to WLAN",
-                    "time": 1579549024893,
-                }
-            ],
-        }
-        controller.api.session_handler("data")
+        mock_unifi_websocket(
+            data={
+                "meta": {"rc": "ok", "message": "events"},
+                "data": [
+                    {
+                        "datetime": "2020-01-20T19:37:04Z",
+                        "key": aiounifi.events.WIRELESS_CLIENT_CONNECTED,
+                        "msg": "User[11:22:33:44:55:66] has connected to WLAN",
+                        "time": 1579549024893,
+                    }
+                ],
+            },
+        )
 
         assert wireless_clients_mock.assert_called_once
 
 
+async def test_reconnect_method(hass, aioclient_mock, caplog):
+    """Verify reconnect prints only when explicitly told to do so."""
+    config_entry = await setup_unifi_integration(hass, aioclient_mock)
+    controller = hass.data[UNIFI_DOMAIN][config_entry.entry_id]
+
+    caplog.clear()
+    with patch.object(controller, "async_reconnect") as mock_reconnect:
```

Comment:
If we patch the aiounifi controller class we should be able to check calls to that patched instance eg the `login` method.